sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala

```diff
@@ -135,33 +125,47 @@ class EquivalentExpressions {
   // For some special expressions we cannot just recurse into all of its children, but we can
   // recursively add the common expressions shared between all of its children.
   private def commonChildrenToRecurse(expr: Expression): Seq[Seq[Expression]] = expr match {
+    case _: CodegenFallback => Nil
     case i: If => Seq(Seq(i.trueValue, i.falseValue))
     case c: CaseWhen =>
       // We look at subexpressions in conditions and values of `CaseWhen` separately. It is
       // because a subexpression in conditions will be run no matter which condition is matched
       // if it is shared among conditions, but it doesn't need to be shared in values. Similarly,
       // a subexpression among values doesn't need to be in conditions because no matter which
       // condition is true, it will be evaluated.
-      val conditions = c.branches.tail.map(_._1)
+      val conditions = if (c.branches.length > 1) {
```

Comment:
This fixes https://github.com/apache/spark/pull/30245#discussion_r660870034

Reply:
@Kimahriman I think this fix works? The only drawback is, if there are common subexpressions among the conditions, they will always be counted as "appear twice" and gets codegened into methods.

I think the perf overhead is really small, and if the first condition is false, we evaluate the next condition which gives perf improvement because of common subexpressions elimination.

For the value branches of `CaseWhen`, I don't touch them in this PR.

Reply:
Yeah this definitely fixes a potential bug of creating subexpressions for things that are never evaluated, same with the coalesce update. I think the values are already handled fine, it's just the conditionals that had an issue with short circuiting

Reply:
This fixed https://github.com/apache/spark/pull/30245#discussion_r660870034. 

> The only drawback is, if there are common subexpressions among the conditions, they will always be counted as "appear twice" and gets codegened into methods.

I just don't get this. You mean for `If(a + b > 1, 1, a + b + c > 1, 2, a + b + c > 2, 3)`, `a + b + c` will be counted twice and considered as common subexpression?

Reply:
I think he means in `CaseWhen(a + b > 1, 1, a + b + c > 1, 2)`, `a + b` will be a subexpression even though it might only be executed once.

Reply:
But `CaseWhen(a + b > 1, 1, a + b + c > 1, 2, a + b + c > 0, 3)`, `a + b + c` won't even be considered for a subexpression if it's seen elsewhere

Reply:
Yes, because the first condition of `CaseWhen` is in both `childrenToRecurse` and `commonChildrenToRecurse`